example/dictionary.sw

```diff
@@ -0,0 +1,440 @@
+end 
+
+def safeGetL : Int -> List a -> Unit + a = \i.\l.
+  case l (\_. inl ()) (\n. if (i <= 0) {inr $ fst n} {safeGetL (i - 1) (snd n)})
+end 
+
+def getL : Int -> List a -> a = \i.\l.
+  case (safeGetL i l) (\_. fail "GET_L OUT OF BOUNDS!") (\a. a)
+end
+
+def formatL : List a -> Text = \l.
+ let f : List a -> Text = \l. case l (\_. "]") (\n. ", " ++ format (fst n) ++ f (snd n))
+ in case l (\_. "[]") (\n. "[" ++ format (fst n) ++ f (snd n))
+end
+
+def inorder : RBTree k -> List k = \t.
+  case t (\_. inl ()) (\n.
+    appendL (inorder n.l) (appendL (pureL n.k) $ inorder n.r)
+  )
+end
+
+/*
+balance B (T R (T R a x0 b) x1 c) x2 d = T R (T B a x0 b) x1 (T B c x2 d) -- case 1: LL red
+balance B (T R a x0 (T R b x1 c)) x2 d = T R (T B a x0 b) x1 (T B c x2 d) -- case 2: LR red
+balance B a x0 (T R (T R b x1 c) x2 d) = T R (T B a x0 b) x1 (T B c x2 d) -- case 3: RL red
+balance B a x0 (T R b x1 (T R c x2 d)) = T R (T B a x0 b) x1 (T B c x2 d) -- case 4: RR red
+balance color a x b = T color a x b -- case 5
+*/
+def balanceT : RBNode k -> RBTree k = \t.
+  let balanced : [ a: RBTree k, x0: k, b: RBTree k,   x1: k,   c: RBTree k, x2: k, d: RBTree k] -> RBTree k
+    = \v.
+    inr [c=red, l=inr [c=black, l=v.a, k=v.x0, r=v.b], k=v.x1, r=inr [c=black, l=v.c, k=v.x2, r=v.d]]
+  in let case5 : RBNode k -> RBTree k =
+    inr  // just id - TODO: inline?
+  in let balanceRR : RBNode k -> RBNode k -> RBTree k = \t.\rn.
+    case rn.r (\_. case5 t) (\rrn.
+      if (rrn.c == red) {
+        balanced [a=t.l, x0=t.k, b=rn.l, x1=rn.k, c=rrn.l, x2=rrn.k, d=rrn.r]
+      } {
+        case5 t
+      }
+    )
+  in let balanceRL : RBNode k -> RBNode k -> RBTree k = \t.\rn.
+    case rn.l (\_. balanceRR t rn) (\rln.
+      if (rln.c == red) {
+        balanced [a=t.l, x0=t.k, b=rln.r, x1=rln.k, c=rln.r, x2=rn.k, d=rn.r]
+      } {
+        balanceRR t rn
+      }
+    )
+  in let balanceR : RBNode k -> RBTree k = \t.
+    case t.r (\_.
+      case5 t
+    ) (\lr.
+      if (lr.c == red) {
+        balanceRL t lr
+      } {
+        case5 t
+      }
+    )
+  in let balanceLR : RBNode k -> RBNode k -> RBTree k = \t.\ln.
+    case ln.r (\_. balanceR t) (\lrn.
+      if (lrn.c == red) {
+        balanced [a=ln.l, x0=ln.k, b=lrn.l, x1=lrn.k, c=lrn.l, x2=t.k, d=t.r]
+      } {
+        balanceR t
+      }
+    )
+  in let balanceLL : RBNode k -> RBNode k -> RBTree k = \t.\ln.
+    case ln.l (\_. balanceLR t ln) (\lln.
+      if (lln.c == red) {
+        balanced [a=lln.l, x0=lln.k, b=lln.r, x1=ln.k, c=ln.r, x2=t.k, d=t.r]
+      } {
+        balanceLR t ln
+      }
+    )
+  in let balanceL : RBNode k -> RBTree k = \t.
+    case t.l (\_.
+      balanceR t
+    ) (\ln.
+      if (ln.c == red) {
+        balanceLL t ln
+      } {
+        balanceR t
+      }
+    )
+  in if (t.c == red) {
+    case5 t
+  } {
+    balanceL t
+  }
+end
+
+def insertT : (k -> a) -> k -> RBTree k -> RBTree k = \p.\x.\t.
```

Comment:
There is a bug somewhere in insert or balance:
```
> def i = insertS end
> def s = i 7 $ i 8 $ i 32 $ i 52 $ i 62 $ i 76 $ i 80 emptyS end
> formatS s
it4 : Text = "[7, 8, 32, 52, 62, 76, 80]"
> formatS $ i 54 s
it5 : Text = "[7, 8, 32, 62, 54, 62, 76, 80]"
```

| S | S + 54 |
|--------|--------|
| <img width="184" alt="Screenshot 2024-10-14 at 10 44 19 PM" src="https://github.com/user-attachments/assets/e328f793-cef1-4ce0-97db-b7ac73aca32e"> | <img width="184" alt="Screenshot 2024-10-14 at 10 44 42 PM" src="https://github.com/user-attachments/assets/969a556b-0114-4cb8-9e44-7e3b9215cafa"> |

Reply:
Fixed! There was a tree duplication in balance in two places caused by copy-paste.